### PR TITLE
Automatically generate contributor lists in documentation and citation file

### LINKED ISSUE #1086: Add individual authors to CITATION file and Generate Authors Webpage from Citation file
As discussed during review of https://github.com/UCL/TLOmodel/pull/1074,  we would like to:

(1) Elaborate the `CITATION.CFF` file to include the names (and ORCID URLS) for each authors, as recommended [here](https://github.com/UCL/TLOmodel/pull/1074).

(2) Make the [authors page ](https://www.tlomodel.org/authors.html) be generated (in part) from the `CITATION.CFF` file.

### PR DESCRIPTION
**Pull request by matt-graham**
Fixes #1086

Adds a top-level `contributors.yaml` file which lists details of contributors to project, using fields from citation file format [`defintions.person` schema](https://github.com/citation-file-format/citation-file-format/blob/main/schema-guide.md#definitionsperson) plus additional custom fields `role`, `contributions` and `github-username` to define respectively role in project, a list of contributions to project and GitHub username if relevant.

A script `docs/tlo_contributors.py` uses this `contributors.yaml` file to generate a file `docs/_contributors_list.html` which is then included in a new `docs/contributors.rst` file which replaces the previous `docs/authors.rst` (we can retain the same file name if wanted, I just thought contributors was a better fit for the range of people included). There is also a script `src/scripts/automation/update_citation.py` which updates the author list in the `CITATION.cff` file from `contributors.yaml`, filtering to people who have `Software development` listed in their `contributions` list field (which I've tried to ensure is everyone who has commits listed on the GitHub repository) - I'm not sure if this is the best way to filter or if any filtering is necessary at all - something to be discussed!

As a first pass I have populated this file from a combination of the existing `docs/authors.rst` documentation source file, the GitHub contributors list and people present in the Slack channel. I have tried to add ORCIDs for anyone I could find one for, though some of these need confirming as the ORCID profiles didn't have enough information to match on anything but the name. In some cases peoples affiliations also need checking in particular if they have multiple affiliations (which CFF doesn't have inbuilt support for) or have moved on from the project and potentially now have no or a different affiliation. I've tried to add `# to check` comments to any fields that I think need verifying, though there will probably also be other mistakes that need fixing - apologies for any errors and/or if I've missed anyone out that should have been included.

For the `contributions` fields I generally followed the categories in the existing authors page in the documentation (`Epidemiology and modelling`, `Software development`, `Clinical consultant`) adding additional categories for `Health economics` and `Project management`. I think the `modelling` bit in `Epidemiology and modelling` may have already been meant to include wider topics like health economics side of model, so this additional category might not be necessary, and/or existing assignments might need reviewing. Currently the first entry in the `contributions` list defines which heading a person is listed under in the generated documentation contributors page, but additional entries can also be defined (for example most people with `Epidemiology and modelling` contribution also have `Software development` as a secondary contribution). If no specific contributions are listed people are added to a catch-all `Acknowledgements` section in contributors documentation page (which currently corresponds to all people currently listed in [acknowledgements section in existing authors documentation page](https://www.tlomodel.org/authors.html#acknowledgements)).

I've also had a go at adding a GitHub Actions workflow which will automatically run the `src/scripts/automation/update_citation.py` script whenever there is a push to `master` which updates `contributors.yaml` and then syncs the changes to `CITATION.cff` if any back to the repository - this might not be necessary / a bit overcomplicating things so can be removed if we think overkill.

Current the author list generated in the `CITATION.cff` file is in the same (arbitrary) order as people are listed in `contributors.yaml`. We could potentially either sort on one of the keys (for example alphabetically by family name) or add an explicit `order` field to sort by.

### CODE REVIEW

File: contributors.yaml [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+  orcid: "https://orcid.org/0000-0003-4275-8509"  # to check
+  affiliation: "Imperial College London" # to check
+  website: https://www.imperial.ac.uk/people/b.jewell
+  github-username: brittajewell
+  contributions:
+    - Epidemiology and modelling
+    - Software development
+- given-names: Ines
+  family-names: Lin
+  orcid: "https://orcid.org/0000-0002-5213-3281"  # to check
```
Comment (matt-graham) [inline contributors.yaml:76]: @inesll can I check if this a correct ORCID for you?

File: contributors.yaml [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+  affiliation: "University College London"
+  github-username: inesll
+  contributions:
+    - Epidemiology and modelling
+    - Software development
+- given-names: Tara
+  family-names: Mangal
+  orcid: "https://orcid.org/0000-0001-9222-8632"  # to check
+  affiliation: "Imperial College London / University of York"  # to check
```
Comment (matt-graham) [inline contributors.yaml:85]: @tdm32 can I check if this is a correct ORCID for you and if you'd like to just be listed with a single affiliation (and if so which!).

File: contributors.yaml [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+  family-names: Graham
+  orcid: "https://orcid.org/0000-0003-0058-263X"
+  affiliation: "University College London"
+  website: "https://willgraham01.github.io/"
+  github-username: willGraham01
+  contributions:
+    - Software development
+- given-names: Margherita
+  family-names: Molaro
+  orcid: "https://orcid.org/0000-0002-1841-4274"  # to check
```
Comment (matt-graham) [inline contributors.yaml:190]: @marghe-molaro can I check if this is a correct ORCID for you?

File: contributors.yaml [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+  affiliation: "Imperial College London"
+  github-username: marghe-molaro
+  contributions:
+    - Epidemiology and modelling
+    - Software development
+- given-names: Eva
+  family-names: Janoušková
+  orcid: "https://orcid.org/0000-0002-4104-0119"  # to check
```
Comment (matt-graham) [inline contributors.yaml:198]: @EvaJanouskova can I check if this a correct ORCID for you?

File: contributors.yaml [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+  affiliation: "Imperial College London"
+  website: "https://www.imperial.ac.uk/people/i.hawryluk19"
+  github-username: ihwaryluk
+  contributions:
+    - Epidemiology and modelling
+    - Software development
+- given-names: Britta
+  family-names: Jewell
+  orcid: "https://orcid.org/0000-0003-4275-8509"  # to check
+  affiliation: "Imperial College London" # to check
```
Comment (matt-graham) [inline contributors.yaml:68]: @brittajewell - not sure if you are still getting GitHub notifications but if so can I check if this is a valid ORCID for you and if the affiliation should be updated?

File: contributors.yaml [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+  website: "https://www.homepages.ucl.ac.uk/~ucbtaut/"
+  github-username: tamuri
+  role: "Lead developer"
+  contributions:
+    - Software development
+- given-names: Joseph
+  family-names: Mfutso-Bengo
+  orcid: "https://orcid.org/0000-0001-9396-3700"
+  affiliation: "Kamuzu University of Health Sciences"
+  website: "https://sphfm.medcol.mw/staff/dr-joseph-mfutso-bengo-phd" # to check - link seems not to work
```
Comment (matt-graham) [inline contributors.yaml:34]:
```suggestion
  website: "https://www.researchgate.net/profile/Joseph-Mfutso-Bengo"
```

As current link (taken from existing authors page) does not seem to work possibly could use ResearchGate page as a replacement?

File: contributors.yaml [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+- given-names: Kate
+  family-names: Bilsborrow
+  affiliation: "Imperial College London"
+  website: "https://www.imperial.ac.uk/people/k.bilsborrow"
+  github-username: imperialkate
+  contributions:
+    - Project management
+- given-names: Newton
+  family-names: Chagoma
+  orcid: "https://orcid.org/0000-0003-0752-1696"  # to check
```
Comment (matt-graham) [inline contributors.yaml:236]: @nchagoma503 can I check if this is a correct ORCID for you?

File: contributors.yaml (hunk 4 above)
Comment (EvaJanouskova) [reply contributors.yaml:198]: Yes, it is correct.

File: contributors.yaml (hunk 1 above)
Comment (inesll) [reply contributors.yaml:76]: Hi @matt-graham! No, sorry, my orcid is https://orcid.org/0000-0002-1945-3208
+ family-names (on my passport): Li Lin

Thank you! :)

**Comment by tamuri** [comment]
> There is also a script `src/scripts/automation/update_citation.py` which updates the author list in the `CITATION.cff` file from `contributors.yaml`, filtering to people who have `Software development` listed in their `contributions` list field (which I've tried to ensure is everyone who has commits listed on the GitHub repository) - I'm not sure if this is the best way to filter or if any filtering is necessary at all - something to be discussed!

We had quite some discussion about this at the TLM kick-off. IRC, the group/consortium-style authorship was preferred. @tbhallett?

File: .github/workflows/update-citation.yml [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+        with:
+          args: "--validate"
+      - name: Push updated citation file back to repository
+        run:
+          git config --local user.name github-actions[bot]
+          git config --local user.email "github-actions[bot]@users.noreply.github.com"
+          if ! git diff --exit-code; then
+            git add CITATION.cff
+            git commit -m "Updating CITATION.cff file"
+            git push -f origin master
```
Comment (tamuri) [inline .github/workflows/update-citation.yml:36]: 1.  master is protected branch. Do we need to update the protections somehow? Here's a [_long_ discussion](https://github.com/orgs/community/discussions/25305).
2.  Why does it need `-f`?

**Comment by tbhallett** [comment]
> > There is also a script `src/scripts/automation/update_citation.py` which updates the author list in the `CITATION.cff` file from `contributors.yaml`, filtering to people who have `Software development` listed in their `contributions` list field (which I've tried to ensure is everyone who has commits listed on the GitHub repository) - I'm not sure if this is the best way to filter or if any filtering is necessary at all - something to be discussed!
> 
> 
> 
> We had quite some discussion about this at the TLM kick-off. IIRC, the group/consortium-style authorship was preferred. @tbhallett?

I'm happy with the recognising the individual contributors in this way. (I believe the consortium style authorship will be used for some research papers and especially ones that do routine analayses of data).

File: .github/workflows/update-citation.yml (hunk 8 above)
Comment (matt-graham) [reply .github/workflows/update-citation.yml:36]: The force push shouldn't be required here I was just not careful enough in checking when I adapted code in [this answer from another discussion](https://github.com/orgs/community/discussions/25234#discussioncomment-4026272).

Yeah branch protections could be problematic here, and thinking about it automatically creating a pull request may be a more sensible approach, or just adding a step to the checks which runs update script and checks if any changes produced (or set up a corresponding pre-commit hook).

Comment (matt-graham) [reply .github/workflows/update-citation.yml]: I've now removed this workflow in favour of just adding an option to just check for differences when running `update_citation.py` script and added this to the `tox` `check` environment. This seems a safer solution and still should achieve the goal of making sure citation file is kept up to date if contributor data is changed.

File: contributors.yaml [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+  orcid: "https://orcid.org/0000-0002-1945-3208"
+  affiliation: "University College London"
+  github-username: inesll
+  contributions:
+    - Epidemiology and modelling
+    - Software development
+- given-names: Tara
+  family-names: Mangal
+  orcid: "https://orcid.org/0000-0001-9222-8632"  # to check
+  affiliation: "Imperial College London / University of York"  # to check
```
Comment (matt-graham) [inline contributors.yaml:85]:
```suggestion
  orcid: "https://orcid.org/0000-0001-9222-8632"
  affiliation: "Imperial College London"
```
Confirmed above ORCID and affliation with Tara

File: contributors.yaml [hunk 10]
```diff
@@ -0,0 +1,276 @@
...
+  contributions:
+    - Epidemiology and modelling
+    - Health economics
+    - Software development
+- given-names: Wingston
+  family-names: Ng'ambi
+  orcid: "https://orcid.org/0000-0003-3967-905X"
+  affiliation: "Kamuzu University of Health Sciences"
+  website: "https://www.linkedin.com/in/wingston-ng-ambi-7950707b"
+  github-username: winnga # to check
```
Comment (matt-graham) [inline contributors.yaml:114]: @tbhallett Do you happen to know if this is GitHub username is Wingston's (it is listed as having some commits associated with it in the GitHub contributors list and Wingston's name seemed most plausible match).

**Comment by tbhallett** [comment]
I can confirm that this is the right ORCID for Britta

**Comment by tbhallett** [comment]
Yes that will be Wingston.

From: Matt Graham ***@***.***>
Date: Monday, 9 October 2023 at 12:13
To: UCL/TLOmodel ***@***.***>
Cc: Hallett, Timothy B ***@***.***>, Mention ***@***.***>
Subject: Re: [UCL/TLOmodel] Automatically generate contributor lists in documentation and citation file (PR #1128)
This email from ***@***.*** originates from outside Imperial. Do not click on links and attachments unless you recognise the sender. If you trust the sender, add them to your safe senders list<https://spam.ic.ac.uk/SpamConsole/Senders.aspx> to disable email stamping for this address.



@matt-graham commented on this pull request.

________________________________

In contributors.yaml<https://github.com/UCL/TLOmodel/pull/1128#discussion_r1350166509>:

> +- given-names: Wingston

+  family-names: Ng'ambi

+  orcid: "https://orcid.org/0000-0003-3967-905X"

+  affiliation: "Kamuzu University of Health Sciences"

+  website: "https://www.linkedin.com/in/wingston-ng-ambi-7950707b"

+  github-username: winnga # to check

@tbhallett<https://github.com/tbhallett> Do you happen to know if this is GitHub username is Wingston's (it is listed as having some commits associated with it in the GitHub contributors list and Wingston's name seemed most plausible match).

**Review by tamuri — APPROVED** [review]
Looks good. I tested locally and noticed that the adding the css override was causing the theme to break on latest sphinx. I've removed that because the latest rtd theme works fine.

**Pull request merged**